### PR TITLE
buildRustPackage: allow not passing a FOD hash

### LINKED ISSUE #138554: buildRustPackage breaks with keepBuildTree or in nix-shell
### Describe the bug
Adding `nativeBuildInputs = [ keepBuildTree ]` to a derivation built with `buildRustPackage` (or building in nix-shell) breaks:

```
$ nix-build -A spotifyd    # after adding keepBuildTree to nativeBuildInputs
...
Executing cargoSetupPostPatchHook
Validating consistency between /build/source//Cargo.lock and /build/spotifyd-0.3.2-vendor.tar.gz/Cargo.lock
/nix/store/y56fwb7pp4h2m017w80jmmcd4wsm2iff-diffutils-3.7/bin/diff: /build/source//Cargo.lock: No such file or directory
/nix/store/y56fwb7pp4h2m017w80jmmcd4wsm2iff-diffutils-3.7/bin/diff: /build/spotifyd-0.3.2-vendor.tar.gz/Cargo.lock: No such file or directory
ERROR: Missing Cargo.lock from src. Expected to find it at: /build/source//Cargo.lock
Hint: You can use the cargoPatches attribute to add a Cargo.lock manually to the build.
builder for '/nix/store/smjh4g3f626v43m6abcv5pcz9k3g0yii-spotifyd-0.3.2.drv' failed with exit code 1
```

The problem seems to be that [cargo-setup-hook.sh](https://github.com/NixOS/nixpkgs/blob/a54d2e72e282f2bc68c49f82c735cf664244ec75/pkgs/build-support/rust/hooks/cargo-setup-hook.sh) assumes the build will always happen in `$NIX_BUILD_TOP`, but that's not the case when e.g. using `keepBuildTree` (which puts the build in `$out/.build`) or when using nix-shell (which puts the build wherever the user ran it from).

### Expected behavior
I expect buildRustPackage to not break when using `keepBuildTree` (like plain stdenv.mkDerivation) and to work in `nix-shell` (also like stdenv.mkDerivation).

### Additional context
The `nix-shell` case can be worked around with `export NIX_BUILD_TOP=$PWD` before running `genericBuild`. (Perhaps this trick can also be used with keepBuildTree, I didn't try.)

### Metadata
Happens on branches release-21.05 and master (a54d2e72e282f2bc68c49f82c735cf664244ec75 @ 2021-09-19).

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Description of changes

This PR fixes a few problems I recently encountered with `buildRustPackage`:

* Previously `cargoHash`/`cargoSha256` would be required arguments, if both were unset `buildRustPackage` would throw an error and the user would have to set either to an empty string/`lib.fakeHash`. Instead we can simply start fetching the dependencies, just like for example `fetchFromGitHub` would without `sha256`. This allows you to write an expression that "prefetches" the hashes with less boilerplate, which is always nice.
* ~~Fixes `buildRustPackage` breaking either in an `nix-shell` environment or when `keepBuildTree` was used. This happened because of a dependency on `NIX_BUILD_TOP` (which is not very reliable, see https://github.com/NixOS/nixpkgs/issues/189691) that messed up paths to the cargo lockfiles.~~ Moved over to #191251.

For more information take a look at the individual commit messages.

 Tested by regenerating a few FOD hashes by commenting out `cargoSha256`.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by IvarWithoutBones** [comment]
Note that this will currently break two packages, which for some reason depend on `cargoDepsCopy` even though its unset later in the hook. The fix is very simple for both of them (`$NIX_BUILD_TOP/$cargoDepsCopy` -> `$cargoDepsCopy`), but im not sure if its preferable to do that in another PR.

* pkgs/applications/blockchains/zcash/default.nix
* pkgs/tools/misc/popsicle/default.nix

**Comment by yu-re-ka** [comment]
`s/buildRustModule/buildRustPackage/` ?

**Comment by IvarWithoutBones** [comment]
> `s/buildRustModule/buildRustPackage/` ?

Oops, looks like my `buildDotnetModule` habits got to me 😅

**Comment by zowoq** [comment]
If you want to split this PR the two hash commits should be fine for master but the NIX_BUILD_TOP commit needs to go to staging.

**Comment by IvarWithoutBones** [comment]
> If you want to split this PR the two hash commits should be fine for master but the NIX_BUILD_TOP commit needs to go to staging.

Done, moved over the `NIX_BUILD_TOP` commit to #191251 👍

File: pkgs/build-support/rust/fetch-cargo-tarball/default.nix [hunk 1]
```diff
@@ -30,7 +30,9 @@ in
 let hash_ =
   if args ? hash then { outputHashAlgo = null; outputHash = args.hash; }
   else if args ? sha256 then { outputHashAlgo = "sha256"; outputHash = args.sha256; }
-  else throw "fetchCargoTarball requires a hash for ${name}";
+  # Prints 'warning: empty hash found, assuming <lib.fakeHash>', generates the FOD hash, and exits afterwards.
+  # This is useful because it allows you to generate the output hash without any boilerplate.
```
Comment (amjoseph-nixpkgs) [inline pkgs/build-support/rust/fetch-cargo-tarball/default.nix:34]: Technically we're not supposed to do this:

https://github.com/NixOS/nixpkgs/blob/ddb18ca71d12ea4edfebe7a9241419ab6fd4f2e7/doc/contributing/coding-conventions.chapter.md?plain=1#L503

IMHO we should instead add some kind of `require-tls` option, but that probably has to be threaded down through nixpkgs and into nix itself.

Comment (amjoseph-nixpkgs) [inline pkgs/build-support/rust/fetch-cargo-tarball/default.nix:34]:
```suggestion
```

Technically we're not supposed to do this.  Everybody does though, but we probably shouldn't *encourage* it.

https://github.com/NixOS/nixpkgs/blob/ddb18ca71d12ea4edfebe7a9241419ab6fd4f2e7/doc/contributing/coding-conventions.chapter.md?plain=1#L503

IMHO we should instead add some kind of `require-tls` option, but that probably has to be threaded down through nixpkgs and into nix itself.

Comment (amjoseph-nixpkgs) [inline pkgs/build-support/rust/fetch-cargo-tarball/default.nix:34]:
```suggestion
```

Technically we're not supposed to do this.  Everybody does though, but we probably shouldn't *encourage* it.

https://github.com/NixOS/nixpkgs/blob/ddb18ca71d12ea4edfebe7a9241419ab6fd4f2e7/doc/contributing/coding-conventions.chapter.md?plain=1#L503

IMHO we should instead add some kind of `require-tls` option, but that probably has to be threaded down through nixpkgs and into nix itself.

**Review by ghost — APPROVED** [review]
This is really useful.  I have something similar in my local tree (plus the ability to pass in the result of `importCargoLock`) for hacking purposes.

File: pkgs/build-support/rust/fetch-cargo-tarball/default.nix (hunk 1 above)
Comment (amjoseph-nixpkgs) [reply pkgs/build-support/rust/fetch-cargo-tarball/default.nix:34]: Also please update the manual:

https://github.com/NixOS/nixpkgs/blob/ddb18ca71d12ea4edfebe7a9241419ab6fd4f2e7/doc/languages-frameworks/rust.section.md?plain=1#L64

The linked paragraph, the paragraph after it, and the sentence before it can be deleted.  The sentence before *that* might require a wording change.

Comment (IvarWithoutBones) [reply pkgs/build-support/rust/fetch-cargo-tarball/default.nix:34]: I've updated the manual, let me know what you think.

**Review by ghost — APPROVED** [review]

File: pkgs/build-support/rust/build-rust-package/default.nix [hunk 2]
```diff
@@ -52,8 +52,6 @@
 , buildAndTestSubdir ? null
 , ... } @ args:
 
-assert cargoVendorDir == null && cargoLock == null -> !(args ? cargoSha256) && !(args ? cargoHash)
-  -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock must be set";
```
Comment (zowoq) [inline pkgs/build-support/rust/build-rust-package/default.nix:56]: > This PR fixes a few problems ...
>
> Previously cargoHash/cargoSha256 would be required arguments, if both were unset buildRustPackage would throw an error and the user would have to set either to an empty string/lib.fakeHash. 

I don't agree that this is a problem. I think the throw is correct and is useful.

> This allows you to write an expression that "prefetches" the hashes with less boilerplate, which is always nice.

I don't see that there is any actual benefit from this.

File: doc/languages-frameworks/rust.section.md [hunk 3]
```diff
@@ -55,20 +55,9 @@ used for [SRI](https://www.w3.org/TR/SRI/) hashes. For example:
...
-```
-
-For `cargoHash` you can use:
-
-```nix
-  cargoHash = lib.fakeHash;
-```
+Cargo hash is obtained by building the package once without
+`cargoSha256` or `cargoHash` set. The correct checksum can then be
+taken from the failed build.
```
Comment (fricklerhandwerk) [inline doc/languages-frameworks/rust.section.md:60]: Why did you remove this? PR [#191378](https://github.com/NixOS/nixpkgs/pull/191378/files#diff-99f4eb67c3def9d78bbb9ec86a7dced190f46ee7151c77dcbb15a6a08727a330R483) seems relevant.

**Comment by AndersonTorres** [comment]
Closing as dead
